File: template/perl/start_server.pl
#!/usr/bin/perl

# start_server.pl
# to run: perl start_server.pl

# start_server: v3

use strict;
use warnings;
use HTTP::Daemon;
use HTTP::Status;
use HTTP::Response;
use File::Spec;
use File::Basename;
use Getopt::Long;
use POSIX qw(strftime);
use Cwd;
use Socket;
use URI::Escape;

my $port = 8000;
my $directory = getcwd();

GetOptions(
	"port|p=i" => \$port,
	"directory|d=s" => \$directory
) or die "Usage: $0 [-p|--port PORT] [-d|--directory DIR]\n";

chdir $directory or die "Cannot change to directory $directory: $!";

sub is_port_in_use {
	my ($port) = @_;
	my $sock;
	my $in_use = 0;

	socket($sock, PF_INET, SOCK_STREAM, getprotobyname('tcp')) or die "socket: $!";
	if (connect($sock, sockaddr_in($port, inet_aton('127.0.0.1')))) {
		$in_use = 1;
	}
	close($sock);

	return $in_use;
}

sub find_available_port {
	my ($start_port) = @_;
	my $port = $start_port;
	while (is_port_in_use($port)) {
		$port++;
	}
	return $port;
}

sub run_server {
	my ($port) = @_;

	my $d = HTTP::Daemon->new(
		LocalAddr => '0.0.0.0',
		LocalPort => $port,
		ReuseAddr => 1,
	) or die "Cannot create server: $!";

	print "Serving HTTP on 0.0.0.0 port $port (http://0.0.0.0:$port/) ...\n";

	while (my $c = $d->accept) {
		while (my $r = $c->get_request) {
			handle_request($c, $r);
		}
		$c->close;
		undef($c);
	}
}

sub handle_request {
	my ($c, $r) = @_;

	if ($r->method eq 'GET') {
		if ($r->uri->path eq '/') {
			serve_file($c, 'index.html');
		} elsif ($r->uri->path eq '/log.html') {
			check_and_generate_report();
			serve_file($c, 'log.html');
		} elsif ($r->uri->path eq '/chat.html') {
			check_and_generate_chat_html();
			serve_file($c, 'chat.html');
		} elsif ($r->uri->path eq '/api/github_update') {
			handle_github_update($c);
		} elsif ($r->uri->path =~ /\.txt$/) {
			serve_text_file($c, $r->uri->path);
		} else {
			serve_file($c, $r->uri->path);
		}
	} elsif ($r->method eq 'POST' && $r->uri->path eq '/chat.html') {
		handle_chat_post($c, $r);
	} else {
		$c->send_error(RC_METHOD_NOT_ALLOWED);
	}
}

sub check_and_generate_report {
	my $html_file = File::Spec->catfile($directory, 'log.html');
	if (!-e $html_file || time() - (stat($html_file))[9] > 60) {
		print "$html_file is older than 60 seconds or does not exist. Running log.html.py...\n";
		system('python log.html.py');
	} else {
		print "$html_file is up-to-date.\n";
	}
}

sub check_and_generate_chat_html {
	my $chat_html_file = File::Spec->catfile($directory, 'chat.html');
	if (!-e $chat_html_file || time() - (stat($chat_html_file))[9] > 60) {
		print "$chat_html_file is older than 60 seconds or does not exist. Running chat.html.py...\n";
		system('python chat.html.py');
	} else {
		print "$chat_html_file is up-to-date.\n";
	}
}

sub handle_github_update {
	my ($c) = @_;
	$c->send_response(HTTP::Response->new(RC_OK, 'OK', ['Content-Type' => 'text/html'], "Update triggered successfully"));
	system('python github_update.py');
}

sub handle_chat_post {
	my ($c, $r) = @_;

	my $content = $r->content;
	my %params = map { split /=/, $_, 2 } split /&/, $content;
	my $author = uri_unescape($params{author} || '');
	my $message = uri_unescape($params{message} || '');

	if ($author && $message) {
		save_message($author, $message);
		$c->send_response(HTTP::Response->new(RC_OK, 'OK', ['Content-Type' => 'text/html'],
			"Message saved successfully" . '<meta http-equiv="refresh" content="1;url=chat.html">'));
		system('python commit_files.py message');
		system('python github_update.py');
	} else {
		$c->send_error(RC_BAD_REQUEST, "Bad Request: Missing author or message");
	}
}

sub save_message {
	my ($author, $message) = @_;

	my $today = strftime("%Y-%m-%d", localtime);
	my $dir = File::Spec->catdir($directory, 'message', $today);
	mkdir $dir unless -d $dir;

	my $title = generate_title($message);
	my $filename = "$title.txt";
	my $filepath = File::Spec->catfile($dir, $filename);

	open my $fh, '>', $filepath or die "Cannot open $filepath: $!";
	print $fh "$message\n\nauthor: $author";
	close $fh;
}

sub generate_title {
	my ($message) = @_;
	my @words = split /\s+/, $message;
	my $title = join '_', @words[0..4];
	$title =~ s/[^a-zA-Z0-9_-]//g;
	$title = substr(rand() . rand(), 2, 10) if $title eq '';
	return $title;
}

sub serve_file {
	my ($c, $path) = @_;
	$path =~ s/^\/+//;
	my $file = File::Spec->catfile($directory, $path);

	if (-f $file) {
		open my $fh, '<', $file or die "Cannot open $file: $!";
		my $content = do { local $/; <$fh> };
		close $fh;

		my $content_type = get_content_type($file);
		$c->send_response(HTTP::Response->new(RC_OK, 'OK', ['Content-Type' => $content_type], $content));
	} else {
		$c->send_error(RC_NOT_FOUND);
	}
}

sub serve_text_file {
	my ($c, $path) = @_;
	$path =~ s/^\/+//;
	my $file = File::Spec->catfile($directory, $path);

	if (-f $file) {
		open my $fh, '<', $file or die "Cannot open $file: $!";
		my $content = do { local $/; <$fh> };
		close $fh;

		my $html_content = generate_html_for_text_file(basename($file), $content);
		$c->send_response(HTTP::Response->new(RC_OK, 'OK', ['Content-Type' => 'text/html; charset=utf-8'], $html_content));
	} else {
		$c->send_error(RC_NOT_FOUND);
	}
}

sub generate_html_for_text_file {
	my ($filename, $content) = @_;
	$content =~ s/&/&amp;/g;
	$content =~ s/</&lt;/g;
	$content =~ s/>/&gt;/g;
	return <<HTML;
<!DOCTYPE html>
<html lang="en">
<head>
	<meta charset="UTF-8">
	<meta name="viewport" content="width=device-width, initial-scale=1.0">
	<title>$filename</title>
	<style>
		body { font-family: Arial, sans-serif; line-height: 1.6; padding: 20px; }
		pre { background-color: #f4f4f4; padding: 15px; border-radius: 5px; white-space: pre-wrap; word-wrap: break-word; }
	</style>
</head>
<body>
	<h1>$filename</h1>
	<pre>$content</pre>
</body>
</html>
HTML
}

sub get_content_type {
	my ($file) = @_;
	my %mime_types = (
		'txt' => 'text/plain',
		'html' => 'text/html',
		'css' => 'text/css',
		'js' => 'application/javascript',
		'json' => 'application/json',
		'png' => 'image/png',
		'jpg' => 'image/jpeg',
		'gif' => 'image/gif',
	);
	my ($ext) = $file =~ /\.([^.]+)$/;
	return $mime_types{lc $ext} || 'application/octet-stream';
}

if (is_port_in_use($port)) {
	print "Port $port is already in use.\n";
	$port = find_available_port($port + 1);
	print "Trying port $port...\n";
}

run_server($port);

# end of start_server.pl

File: template/php/start_server.php
<?php

# start_server.php
# to run: php start_server.php

# start_server: v3

class CustomHTTPRequestHandler {
	private $directory;

	public function __construct($directory = null) {
		$this->directory = $directory ?: getcwd();
	}

	public function handleRequest($uri, $method) {
		if ($uri == '/' || $uri == '') {
			$this->serveFile('/index.html');
		} elseif ($uri == '/log.html') {
			$this->checkAndGenerateReport();
			$this->serveFile('/log.html');
		} elseif ($uri == '/chat.html') {
			$this->checkAndGenerateChatHtml();
			$this->serveFile('/chat.html');
		} elseif ($uri == '/api/github_update') {
			$this->triggerGithubUpdate();
		} elseif (substr($uri, -4) === '.txt') {
			$this->serveTextFile($uri);
		} elseif ($method === 'POST' && $uri == '/chat.html') {
			$this->handleChatPost();
		} else {
			$this->serveFile($uri);
		}
	}

	private function checkAndGenerateReport() {
		$htmlFile = $this->directory . '/log.html';
		if (!file_exists($htmlFile) || (time() - filemtime($htmlFile)) > 60) {
			echo "log.html is older than 60 seconds or does not exist. Running log.html.py...\n";
			exec('python log.html.py');
		} else {
			echo "log.html is up-to-date.\n";
		}
	}

	private function checkAndGenerateChatHtml() {
		$chatHtmlFile = $this->directory . '/chat.html';
		if (!file_exists($chatHtmlFile) || (time() - filemtime($chatHtmlFile)) > 60) {
			echo "chat.html is older than 60 seconds or does not exist. Running chat.html.py...\n";
			exec('python chat.html.py');
		} else {
			echo "chat.html is up-to-date.\n";
		}
	}

	private function triggerGithubUpdate() {
		header('Content-Type: text/html');
		echo "Update triggered successfully";
		exec('python github_update.py');
	}

	private function handleChatPost() {
		$author = $_POST['author'] ?? '';
		$message = $_POST['message'] ?? '';

		if ($author && $message) {
			$this->saveMessage($author, $message);
			header('Content-Type: text/html');
			echo "Message saved successfully";
			echo '<meta http-equiv="refresh" content="1;url=/chat.html">';
			exec('python commit_files.py message');
			exec('python github_update.py');
		} else {
			header("HTTP/1.0 400 Bad Request");
			echo "Bad Request: Missing author or message";
		}
	}

	private function saveMessage($author, $message) {
		$today = date('Y-m-d');
		$directory = $this->directory . "/message/$today";
		if (!is_dir($directory)) {
			mkdir($directory, 0777, true);
		}

		$title = $this->generateTitle($message);
		$filename = "$title.txt";
		$filepath = "$directory/$filename";

		file_put_contents($filepath, $message . "\n\nauthor: $author");
	}

	private function generateTitle($message) {
		$words = array_slice(explode(' ', $message), 0, 5);
		$title = implode('_', $words);
		$title = preg_replace('/[^a-zA-Z0-9_-]/', '', $title);
		if (empty($title)) {
			$title = substr(str_shuffle('abcdefghijklmnopqrstuvwxyz'), 0, 10);
		}
		return $title;
	}

	private function serveFile($uri) {
		$filePath = $this->directory . $uri;
		if (file_exists($filePath) && !is_dir($filePath)) {
			$mimeType = $this->getMimeType($filePath);
			header("Content-Type: $mimeType");
			readfile($filePath);
		} else {
			$this->sendNotFound();
		}
	}

	private function serveTextFile($uri) {
		$filePath = $this->directory . $uri;
		if (file_exists($filePath) && !is_dir($filePath)) {
			$content = file_get_contents($filePath);
			$htmlContent = $this->generateHtmlForTextFile(basename($filePath), $content);
			header("Content-Type: text/html; charset=utf-8");
			echo $htmlContent;
		} else {
			$this->sendNotFound();
		}
	}

	private function generateHtmlForTextFile($filename, $content) {
		$escapedContent = htmlspecialchars($content);
		return <<<HTML
	<!DOCTYPE html>
	<html lang="en">
	<head>
	    <meta charset="UTF-8">
	    <meta name="viewport" content="width=device-width, initial-scale=1.0">
	    <title>$filename</title>
	    <style>
		body { font-family: Arial, sans-serif; line-height: 1.6; padding: 20px; }
		pre { background-color: #f4f4f4; padding: 15px; border-radius: 5px; white-space: pre-wrap; word-wrap: break-word; }
	    </style>
	</head>
	<body>
	    <h1>$filename</h1>
	    <pre>$escapedContent</pre>
	</body>
	</html>
	HTML;
	}

	private function sendNotFound() {
		header("HTTP/1.0 404 Not Found");
		echo "404 Not Found";
	}

	private function getMimeType($filePath) {
		$mimeTypes = [
			'txt' => 'text/plain',
			'html' => 'text/html',
			'css' => 'text/css',
			'js' => 'application/javascript',
			'json' => 'application/json',
			'png' => 'image/png',
			'jpg' => 'image/jpeg',
			'gif' => 'image/gif',
		];

		$extension = strtolower(pathinfo($filePath, PATHINFO_EXTENSION));
		return isset($mimeTypes[$extension]) ? $mimeTypes[$extension] : 'application/octet-stream';
	}
}

function isPortInUse($port) {
	$connection = @fsockopen('localhost', $port);
	if (is_resource($connection)) {
		fclose($connection);
		return true;
	}
	return false;
}

function findAvailablePort($startPort) {
	$port = $startPort;
	while (isPortInUse($port)) {
		$port++;
	}
	return $port;
}

function runServer($port, $directory) {
	$host = '0.0.0.0';
	echo "Starting server on http://$host:$port\n";
	echo "Serving directory: $directory\n";

	$command = sprintf(
		'php -S %s:%d -t %s %s',
		$host,
		$port,
		escapeshellarg($directory),
		escapeshellarg(__FILE__)
	);

	passthru($command);
}

// Parse command line arguments
$options = getopt("p:d:", ["port:", "directory:"]);
$port = $options['p'] ?? $options['port'] ?? null;
$directory = $options['d'] ?? $options['directory'] ?? getcwd();

if ($port === null) {
	$port = 8000;
	while (isPortInUse($port)) {
		$port = findAvailablePort($port + 1);
		echo "Trying port $port...\n";
	}
} elseif (isPortInUse($port)) {
	echo "Port $port is already in use.\n";
	exit(1);
}

// Run the server
runServer($port, $directory);

// Handle incoming requests
if (php_sapi_name() === 'cli-server') {
	$handler = new CustomHTTPRequestHandler($directory);
	$handler->handleRequest($_SERVER['REQUEST_URI'], $_SERVER['REQUEST_METHOD']);
}

# end of start_server.php